File: core/workspace/widget_catalog_test.go
package workspace

import "testing"

func TestWidgetKindCatalogIsStableAndHonest(t *testing.T) {
	t.Parallel()

	catalog := WidgetKindCatalog()
	expectedKinds := []WidgetKind{
		WidgetKindTerminal,
		WidgetKindFiles,
		WidgetKindCommander,
		WidgetKindPreview,
		WidgetKindEditor,
		WidgetKindWeb,
	}
	if len(catalog) != len(expectedKinds) {
		t.Fatalf("expected %d widget kinds, got %#v", len(expectedKinds), catalog)
	}

	seen := make(map[WidgetKind]struct{}, len(catalog))
	for index, expectedKind := range expectedKinds {
		entry := catalog[index]
		if entry.Kind != expectedKind {
			t.Fatalf("unexpected kind at index %d: %#v", index, catalog)
		}
		if entry.Label == "" || entry.Description == "" || entry.DefaultTitle == "" {
			t.Fatalf("catalog entry must be operator-readable: %#v", entry)
		}
		if _, exists := seen[entry.Kind]; exists {
			t.Fatalf("duplicate widget kind in catalog: %s", entry.Kind)
		}
		seen[entry.Kind] = struct{}{}
	}

	terminal := catalog[0]
	if terminal.Status != WidgetKindStatusAvailable || !terminal.RuntimeOwned || !terminal.CanCreate {
		t.Fatalf("terminal must remain an available backend-owned creatable widget kind: %#v", terminal)
	}

	commander := catalog[2]
	if commander.Status != WidgetKindStatusFrontendLocal || commander.RuntimeOwned || commander.CanCreate {
		t.Fatalf("commander must not be overclaimed as backend-owned yet: %#v", commander)
	}

	for _, entry := range catalog[3:] {
		if entry.Status != WidgetKindStatusPlanned || entry.RuntimeOwned || entry.CanCreate {
			t.Fatalf("future widget kind must remain planned until implemented: %#v", entry)
		}
	}
}

func TestWidgetKindCatalogReturnsCopy(t *testing.T) {
	t.Parallel()

	catalog := WidgetKindCatalog()
	catalog[0].Kind = "mutated"

	if WidgetKindCatalog()[0].Kind != WidgetKindTerminal {
		t.Fatalf("catalog mutation leaked into shared state")
	}
}

File: frontend/src/features/workspace/model/widget-catalog.test.ts
import { describe, expect, it } from 'vitest'

import { type WorkspaceWidgetKindCatalogEntry } from '@/shared/api/workspace'
import {
  getWorkspaceWidgetKindEntry,
  isWorkspaceWidgetKindCreatable,
  isWorkspaceWidgetKindFrontendLocal,
} from './widget-catalog'

const catalog: WorkspaceWidgetKindCatalogEntry[] = [
  {
    kind: 'terminal',
    label: 'Terminal',
    description: 'Backend terminal',
    status: 'available',
    runtime_owned: true,
    can_create: true,
    supports_connections: true,
    supports_path: false,
    default_title: 'Terminal',
  },
  {
    kind: 'commander',
    label: 'Commander',
    description: 'Frontend local commander',
    status: 'frontend-local',
    runtime_owned: false,
    can_create: false,
    supports_connections: false,
    supports_path: false,
    default_title: 'Commander',
  },
  {
    kind: 'preview',
    label: 'Preview',
    description: 'Future preview',
    status: 'planned',
    runtime_owned: false,
    can_create: false,
    supports_connections: false,
    supports_path: true,
    default_title: 'Preview',
  },
]

describe('workspace widget catalog helpers', () => {
  it('resolves backend-creatable kinds only from available runtime-owned entries', () => {
    expect(isWorkspaceWidgetKindCreatable(catalog, 'terminal')).toBe(true)
    expect(isWorkspaceWidgetKindCreatable(catalog, 'commander')).toBe(false)
    expect(isWorkspaceWidgetKindCreatable(catalog, 'preview')).toBe(false)
    expect(isWorkspaceWidgetKindCreatable(catalog, 'missing')).toBe(false)
  })

  it('distinguishes frontend-local kinds from planned or backend-owned kinds', () => {
    expect(isWorkspaceWidgetKindFrontendLocal(catalog, 'commander')).toBe(true)
    expect(isWorkspaceWidgetKindFrontendLocal(catalog, 'terminal')).toBe(false)
    expect(isWorkspaceWidgetKindFrontendLocal(catalog, 'preview')).toBe(false)
  })

  it('returns undefined for unknown catalog kinds', () => {
    expect(getWorkspaceWidgetKindEntry(catalog, 'missing')).toBeUndefined()
  })
})
